File: java_hummus/src/main/java/com/github/klee0kai/hummus/collections/weaklist/SoftList.java
package com.github.klee0kai.hummus.collections.weaklist;

import com.github.klee0kai.hummus.model.IProvide;
import com.github.klee0kai.hummus.model.RefProvide;

import java.util.Collection;
import java.util.List;

public class SoftList<T> extends RefList<T> {

    public SoftList() {

    }

    public SoftList(Collection<? extends T> c) {
        this();
        addAll(c);
    }


    @Override
    IProvide<T> wrapRef(T val) {
        return RefProvide.soft(val);
    }

    @Override
    RefList<T> createNew(List<T> list) {
        return new SoftList<>(list);
    }
}

File: hummus/src/jvmMain/kotlin/com/github/klee0kai/hummus/collections/weaklist/SoftList.kt
package com.github.klee0kai.hummus.collections.weaklist

import com.github.klee0kai.hummus.model.IProvide
import com.github.klee0kai.hummus.model.RefProvide

open class SoftList<T> constructor() : RefList<T>() {

    constructor(list: Iterable<T?>) : this() {
        addAll(list)
    }

    override fun wrapRef(it: T?): IProvide<T?> {
        return RefProvide.soft(it)
    }

    override fun createNew(list: List<T?>): RefList<T> {
        return SoftList(list)
    }
}